src/catchup/ApplyCheckpointWork.cpp

```diff
@@ -266,9 +302,22 @@ ApplyCheckpointWork::onRun()
         }
 
         auto result = applyHistoryOfSingleLedger();
-        auto const& lm = mApp.getLedgerManager();
-        auto done = lm.getLastClosedLedgerNum() == mCheckpointRange.mLast;
+        auto lcl = mApp.getLedgerManager().getLastClosedLedgerNum();
+
+        // In case we just closed a ledger that unblocked publishing of the
+        // previous checkpoint, wait for publish to complete before
+        // continuing application
+        if (result && mWaitForPublish)
+        {
+            auto ledger = hm.prevCheckpointLedger(lcl);
+            if (ledger == lcl && hm.publishQueueLength() > 0)
```

Comment:
isn't the condition on the size of the publish queue enough?

Reply:
now that I think of it we should move this code into a precondition and use `ConditionalWork` to wrap `ApplyCheckpointWork`

Reply:
> isn't the condition on the size of the publish queue enough?

The sequence of events is as follows: a ledger that is the last one in a checkpoint is closed, triggering snapshot being queued for publish. Then `ResolveSnapshotWork` waits for the next ledger (aka first ledger in the next checkpoint) to be closed in order to proceed. So the "publish queue is not empty" condition is not enough here, as we need to also close the next ledger to unblock `ResolveSnapshotWork` before going into `WORK_WAITING` state.

> now that I think of it we should move this code into a precondition and use ConditionalWork to wrap ApplyCheckpointWork

Right, I think initially this was a bit tricky to do, since I did not allow any drift in catchup vs publish. But with the min/max queue size to trigger wait/applying that you suggested in the other comment, this should be cleaner to do with ConditionalWork (since we're waiting for checkpoint far enough in the past to complete publishing, so I _think_ it's not going to have the `ResolveSnapshotWork` problem I mentioned above)